### DISCUSSION TITLE
High CPU use

### DISCUSSION THREAD

**Discussion by myangelkamikaze** [cefsharp/CefSharp#3892] [General]
I don't really know what's going on here, so this is just a description of what I observed. There's a tl;dr at the end.

Recently I upgraded CefSharp from version 92.0.251 to 94.4.50 and a user noticed very high CPU use (changed from 15% to 91%). Version 95.7.141 had the same problem.
![image](https://user-images.githubusercontent.com/40002167/142569868-36970459-4a84-4ceb-beea-a6ca58a0fa9c.png)
![image](https://user-images.githubusercontent.com/40002167/142569877-baafbb51-46da-41fd-8b34-dbf1b950ea72.png)

After a bit I figured that hardware acceleration isn't enabled, however it was enabled. CPU/GPU
![image](https://user-images.githubusercontent.com/40002167/142569583-23e257d1-c94a-4041-8da7-3cb27f7a27c9.png)
![image](https://user-images.githubusercontent.com/40002167/142569592-43d0dfce-08a7-426e-91e8-4cc49a7cc155.png)

I couldn't reproduce the behavior with hardware acceleration enabled, but I had the same CPU use change with hardware acceleration disabled. After searching for a while, I noticed this comment: https://github.com/cefsharp/CefSharp/issues/3817#issuecomment-931760480. I noticed that versions 92 and 95 were using different --use-gl parameters, so I tried to change that, but that didn't get applied.

Since I couldn't get a fix with command line arguments, I eventually just deleted the files mentioned in this comment: https://github.com/cefsharp/CefSharp/issues/3817#issuecomment-948117902. Without those files, the CPU use went back to normal.

My conclusion is that the vulkan implementation will simply kill CPU performance on older hardware, so as a workaround, I'm deleting those files. This workaround doesn't seem ideal, so I'm wondering if there's a way to disable the vulkan software rendering somehow.

Extra: the testing was happening in this [twitter thread](https://twitter.com/apple_9119/status/1459383480232538113), I'm linking it in case I forgot anything.
Thread notes:
First tweet is just pointing out the high CPU use.
The image where I'm asking if hardware acceleration is enabled just shows where the option is.
The in game picture is meant to confirm hardware acceleration works, in case it didn't the HP bar would look like this: ![image](https://user-images.githubusercontent.com/40002167/142572709-26614c15-31ac-44aa-a501-0e93fbab6291.png)

tl;dr
The vulkan software rendering can cause very high CPU use on older hardware, is there a way to disable it?

**Comment by amaitland** [comment]
What did you try exactly? The links I provided in #3817 Have all the information that I'm aware of. I haven't had need to look for additional information. 

Chromium decides which rendering implementation to use, you can load chrome://gpu in Chrome or a ChromiumWebBrowser instance to check the GPU status at runtime. 

Swiftshader will be removed at some point in the future, so any workaround will only be temporary. I'd suggest attempting to reproduce the issue with chrome and file a bug on the Chromium issue tracker (search for existing first).

**Reply by myangelkamikaze** [reply]
What exactly, as in command line arguments?

I tried both --use-gl=swiftshader and --use-gl=swiftshader-webgl (and a few others, but reading the links again that was probably pointless). Both of those result in the same behavior:
![image](https://user-images.githubusercontent.com/40002167/142973337-5d63ffe4-9c2f-4a84-ae71-e9ea7b9021ce.png)
![image](https://user-images.githubusercontent.com/40002167/142973377-0db133f4-bcfd-4ab9-be0e-87e15e867c20.png)

(should I just copy the whole report from chrome://gpu?)

There are 9 CefSharp.BrowserSubProcess processes. 2 of them have the correct use-gl, 6 don't have the flag and the high CPU use one has this:
![image](https://user-images.githubusercontent.com/40002167/142973737-b833d2ce-2609-4f64-a923-1d0965f08254.png)

So the flag gets set correctly, but one of the processes ignores it?

I'll try to make a minimal reproduction. Does chrome in this case have to be googles release or is any up to date chromium browser, like edge, fine for it?

**Reply by amaitland** [reply]
It's possible the command line args only worked in a previous version, I've not really done any testing with newer versions.

> There are 9 CefSharp.BrowserSubProcess processes. 2 of them have the correct use-gl, 6 don't have the flag and the high CPU use one has this:

If `--type=gpu-process` has the command line arg set then I would expect that it's passed through correctly.

> I'll try to make a minimal reproduction. Does chrome in this case have to be googles release or is any up to date chromium browser, like edge, fine for it?

I don't know for sure, likely `Edge Chromium`  would be fine.
